### PR TITLE
Throw meaningful exception when DbContext is already registered

### LINKED ISSUE #744: [WebToolsE2E][Aspire]After adding 'Aspire.Npgsql.EntityFrameworkCore.PostgreSQL' component and updating related codes, F5 failed with an exception: 'Some services are not able to be constructed ...'
**INSTALL STEPS**

1. Clean machine: Win11 x64 23h2 ENU
2. Install VS 17.9 P1 from build [34307.189.d17.9] 
3. Apply NuGet Feeds

**REPRO STEPS**

1. In the VS, create a new project > ASP.NET Core Web App(Razor Page) > .NET 8.0 > check 'Enlist in Aspire Orchestration' > Create
2. Right-click the Models folder > Add > Class, named Movie.cs and paste following codes:
```
	public class Movie
	{
	    public int ID { get; set; }
	    public string Title { get; set; } = string.Empty;
	    public DateTime ReleaseDate { get; set; }
	    public string Genre { get; set; } = string.Empty;
	    public decimal Price { get; set; }
	}
```
3. Right-click the Controllers folder > add > Controller > Mvc Controller with views, using Entity Framework:

    - Model class: Movies.cs
    - DbContext class: click '+' to create a new db context
    - Database provider: select PstgreSQL

4. Right-click the razor project > Add > Aspire component > Aspire.Npgsql.EntityFrameworkCore.PostgreSQL using matched version.
5. Open appsettings.json file under Razor project, type following codes:	
```
"Aspire": {
	    "Npgsql": {
	      "EntityFrameworkCore": {
	        "PostgreSQL": {
	          "DbContextPooling": true,
	          "HealthChecks": false,
	          "Tracing": false
	        }
	      }
	    }
	  }
```
6. Then open Program.cs of razor project, type following codes and replace with the DbContext created in step 3:
    `builder.AddNpgsqlDbContext<MyDbContext>("myConnection");`
7. Open appsettings.json of xxx.AppHost project, add following codes and replace with the connection string created in step 3:
```
"ConnectionStrings": {
	        "myConnection": "Host=myserver;Database=test"
	      }
```
8. Open Program.cs of xxx.AppHost project, add following codes:
```
var postgresdb = builder.AddPostgresContainer("pg").AddDatabase("postgresdb");

var myService = builder.AddProject<Projects.MyService>()
                       .WithReference(postgresdb);
```
9. F5

**ACTUAL**
F5 failed with an exception
image
System.AggregateException
  HResult=0x80131500
  Message=Some services are not able to be constructed (Error while validating the service descriptor 'ServiceType: Microsoft.EntityFrameworkCore.Internal.IDbContextPool`1[WebApplication4.Data.WebApplication4Context] Lifetime: Singleton ImplementationType: Microsoft.EntityFrameworkCore.Internal.DbContextPool`1[WebApplication4.Data.WebApplication4Context]': Cannot consume scoped service 'Microsoft.EntityFrameworkCore.DbContextOptions`1[WebApplication4.Data.WebApplication4Context]' from singleton 'Microsoft.EntityFrameworkCore.Internal.IDbContextPool`1[WebApplication4.Data.WebApplication4Context]'.) (Error while validating the service descriptor 'ServiceType: Microsoft.EntityFrameworkCore.Internal.IScopedDbContextLease`1[WebApplication4.Data.WebApplication4Context] Lifetime: Scoped ImplementationType: Microsoft.EntityFrameworkCore.Internal.ScopedDbContextLease`1[WebApplication4.Data.WebApplication4Context]': Cannot consume scoped service 'Microsoft.EntityFrameworkCore.DbContextOptions`1[WebApplication4.Data.WebApplication4Context]' from singleton 'Microsoft.EntityFrameworkCore.Internal.IDbContextPool`1[WebApplication4.Data.WebApplication4Context]'.)
  Source=Microsoft.Extensions.DependencyInjection
  StackTrace:
   at Microsoft.Extensions.DependencyInjection.ServiceProvider..ctor(ICollection`1 serviceDescriptors, ServiceProviderOptions options)
   at Microsoft.Extensions.DependencyInjection.ServiceCollectionContainerBuilderExtensions.BuildServiceProvider(IServiceCollection services, ServiceProviderOptions options)
   at Microsoft.Extensions.Hosting.HostApplicationBuilder.Build()
   at Microsoft.AspNetCore.Builder.WebApplicationBuilder.Build()
   at Program.<Main>$(String[] args) in C:\Users\v-cazhang\source\repos\WebApplication4\WebApplication4\Program.cs:line 14

  This exception was originally thrown at this call stack:
    [External Code]

Inner Exception 1:
InvalidOperationException: Error while validating the service descriptor 'ServiceType: Microsoft.EntityFrameworkCore.Internal.IDbContextPool`1[WebApplication4.Data.WebApplication4Context] Lifetime: Singleton ImplementationType: Microsoft.EntityFrameworkCore.Internal.DbContextPool`1[WebApplication4.Data.WebApplication4Context]': Cannot consume scoped service 'Microsoft.EntityFrameworkCore.DbContextOptions`1[WebApplication4.Data.WebApplication4Context]' from singleton 'Microsoft.EntityFrameworkCore.Internal.IDbContextPool`1[WebApplication4.Data.WebApplication4Context]'.

Inner Exception 2:
InvalidOperationException: Cannot consume scoped service 'Microsoft.EntityFrameworkCore.DbContextOptions`1[WebApplication4.Data.WebApplication4Context]' from singleton 'Microsoft.EntityFrameworkCore.Internal.IDbContextPool`1[WebApplication4.Data.WebApplication4Context]'.

**EXPECTED**
Running the solution succeed and can show the dashboard.

### PR DESCRIPTION
**Pull request by sebastienros**
Fixes #744
 ###### Microsoft Reviewers: [Open in CodeFlow](https://microsoft.github.io/open-pr/?codeflow=https://github.com/dotnet/aspire/pull/2449)

### CODE REVIEW

**Comment by sebastienros** [comment]
Currently adding the `IsDevelopment` check which I forgot

File: src/Components/Common/EntityFrameworkUtils.cs [hunk 1]
```diff
@@ -82,4 +82,14 @@ public static void PatchServiceDescriptor<TContext>(this IHostApplicationBuilder
 
         builder.Services.Add(dbContextOptionsDescriptor);
     }
+
+    public static void EnsureDbContextNotRegistered<TContext>(this IHostApplicationBuilder builder, [CallerMemberName] string callerMemberName = "") where TContext : DbContext
+    {
+        var oldDbContextOptionsDescriptor = builder.Services.FirstOrDefault(sd => sd.ServiceType == typeof(DbContextOptions<TContext>));
+
+        if (oldDbContextOptionsDescriptor is { })
```
Comment (eerhardt) [inline src/Components/Common/EntityFrameworkUtils.cs:90]: (nit) I think this is more readable.

```suggestion
        if (oldDbContextOptionsDescriptor is not null)
```

File: src/Components/Common/EntityFrameworkUtils.cs [hunk 2]
```diff
@@ -82,4 +82,14 @@ public static void PatchServiceDescriptor<TContext>(this IHostApplicationBuilder
...
+        {
+            throw new InvalidOperationException($"DbContext<{typeof(TContext).Name}> is already registered. Please ensure AddDbContext<{typeof(TContext).Name}>() is not invoked before {callerMemberName}().");
```
Comment (eerhardt) [inline src/Components/Common/EntityFrameworkUtils.cs:92]: Should we also recommend the `EnrichXXXDbContext` methods?

**Comment by sebastienros** [comment]
Updated

File: src/Components/Common/EntityFrameworkUtils.cs [hunk 3]
```diff
@@ -82,4 +82,19 @@ public static void PatchServiceDescriptor<TContext>(this IHostApplicationBuilder
...
+        if (!builder.Environment.IsDevelopment())
+        {
+            return;
+        }
+
+        var oldDbContextOptionsDescriptor = builder.Services.FirstOrDefault(sd => sd.ServiceType == typeof(DbContextOptions<TContext>));
+
+        if (oldDbContextOptionsDescriptor is not null)
+        {
+            throw new InvalidOperationException($"DbContext<{typeof(TContext).Name}> is already registered. Please ensure AddDbContext<{typeof(TContext).Name}>() is not invoked before {callerMemberName}() or use the corresponding 'Enrich' method.");
```
Comment (eerhardt) [inline src/Components/Common/EntityFrameworkUtils.cs:97]:
```suggestion
            throw new InvalidOperationException($"DbContext<{typeof(TContext).Name}> is already registered. Please ensure 'services.AddDbContext<{typeof(TContext).Name}>()' is not used when calling '{callerMemberName}()' or use the corresponding 'Enrich' method.");
```

**Comment by eerhardt** [comment]
The tests are broken now. 😦

**Comment by sebastienros** [comment]
@eerhardt was an issue when a PR got merged to main, see https://github.com/dotnet/aspire/pull/2451 for resolution.

**Review by eerhardt — APPROVED** [review]

**Pull request merged**